server/sonar-db-dao/src/main/java/org/sonar/db/issue/IssueDao.java

```diff
@@ -0,0 +1,113 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.db.issue;
+
+import com.google.common.base.Function;
+import com.google.common.base.Predicates;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.sonar.db.Dao;
+import org.sonar.db.DbSession;
+import org.sonar.db.RowNotFoundException;
+
+import static com.google.common.collect.FluentIterable.from;
+import static org.sonar.db.DatabaseUtils.executeLargeInputs;
+
+public class IssueDao implements Dao {
+
+  public java.util.Optional<IssueDto> selectByKey(DbSession session, String key) {
+    return java.util.Optional.ofNullable(mapper(session).selectByKey(key));
+  }
+
+  public IssueDto selectOrFailByKey(DbSession session, String key) {
+    java.util.Optional<IssueDto> issue = selectByKey(session, key);
+    if (!issue.isPresent()) {
+      throw new RowNotFoundException(String.format("Issue with key '%s' does not exist", key));
+    }
+    return issue.get();
+  }
+
+  /**
+   * Gets a list issues by their keys. The result does NOT contain {@code null} values for issues not found, so
+   * the size of result may be less than the number of keys. A single issue is returned
+   * if input keys contain multiple occurrences of a key.
+   * <p>Results may be in a different order as input keys (see {@link #selectByOrderedKeys(DbSession, List)}).</p>
+   */
+  public List<IssueDto> selectByKeys(final DbSession session, List<String> keys) {
+    return executeLargeInputs(keys, mapper(session)::selectByKeys);
+  }
+
+  /**
+   * Gets a list issues by their keys. The result does NOT contain {@code null} values for issues not found, so
+   * the size of result may be less than the number of keys. A single issue is returned
+   * if input keys contain multiple occurrences of a key.
+   * <p>Contrary to {@link #selectByKeys(DbSession, List)}, results are in the same order as input keys.</p>
+   */
+  public List<IssueDto> selectByOrderedKeys(DbSession session, List<String> keys) {
+    List<IssueDto> unordered = selectByKeys(session, keys);
+    return from(keys).transform(new KeyToIssue(unordered)).filter(Predicates.notNull()).toList();
+  }
+
+  private static class KeyToIssue implements Function<String, IssueDto> {
+    private final Map<String, IssueDto> map = new HashMap<>();
+
+    private KeyToIssue(Collection<IssueDto> unordered) {
+      for (IssueDto dto : unordered) {
+        map.put(dto.getKey(), dto);
+      }
+    }
+
+    @Nullable
+    @Override
+    public IssueDto apply(@Nonnull String issueKey) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Remove this "Nonnull" annotation to honor the overridden method's contract. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2638)